src/grab.ts

```diff
@@ -1,157 +1,121 @@
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodUnknown
-      ) {
-        // Split base and conditional fields
-        const conditionalFields = Object.values(
-          conditionalSelections || {}
-        ) as Selection[];
+    // Split base and conditional fields
+    const conditionalFields = Object.values(
+      conditionalSelections || {}
+    ) as Selection[];
 
-        const baseSchema = z.object(toSchemaInput(selection));
-        const conditionalFieldSchemas = conditionalFields.map((field) =>
-          baseSchema.merge(z.object(toSchemaInput(field)))
-        );
-        const s =
-          conditionalFieldSchemas.length === 0
-            ? baseSchema
-            : // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-              // @ts-ignore Need to figure out how to make tsc happy
-              z.union([...conditionalFieldSchemas, baseSchema]);
+    const baseSchema = z.object(toSchemaInput(selection));
+    const conditionalFieldSchemas = conditionalFields.map((field) =>
+      baseSchema.merge(z.object(toSchemaInput(field)))
+    );
+    const s =
+      conditionalFieldSchemas.length === 0
+        ? baseSchema
+        : // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore Need to figure out how to make tsc happy
+          z.union([...conditionalFieldSchemas, baseSchema]);
 
-        return prev.schema instanceof z.ZodArray ? z.array(s) : s;
-      }
-      // If we're already dealing with an object schema inside our array, we need to do a Pick
-      else if (
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodObject
-      ) {
-        const toPick = Object.keys(selection).reduce<{
-          [key: string]: true;
-        }>((acc, key) => {
-          acc[key] = true;
-          return acc;
-        }, {});
+    return schema instanceof z.ZodArray ? z.array(s) : s;
+  })();
 
-        return prev.schema instanceof z.ZodArray
-          ? z.array(prev.schema.element.pick(toPick))
-          : (prev.schema as z.ZodObject<any>).pick(toPick);
-      }
-      // If not unknown/object, I don't know what happened 👀
-      else {
-        return z.never();
-      }
-    })();
+  const res = (newSchema instanceof z.ZodArray
+    ? new ArrayResult({
+        query: query + `{${projections.join(", ")}}`,
+        schema: newSchema,
+      })
+    : new EntityResult({
+        query: query + `{${projections.join(", ")}}`,
+        schema: newSchema,
+      })) as unknown as ArrayResult<AllSelection>;
 
-    return {
-      query: prev.query + `{${projections.join(", ")}}`,
-      schema: schema,
-    } as BaseResult<NewType>;
-  };
+  return res as unknown as NewType;
+};
 
+/**
+ * Misc util
+ */
 type Field<T extends z.ZodType> = T;
 type FromField<T> = T extends Field<infer R>
   ? R
   : T extends [string, infer R]
   ? R
   : z.ZodNever;
```

Comment:
Any reason you use `z.ZodNever` instead of `never`? My understanding of it is that if you've reached this point in the nested ternary, it's an issue with what the dev is passing as a Selector; it doesn't conform to the groqd-accepted syntax of `q.boolean()` or `["active", q.boolean()]`, for example. So the problem is more in groqd-land than it is in zod-land.